### ISSUE TITLE
[Bug]: Training suddenly stops at 25000 timesteps and Optuna optimization immediately exits in my custom environment

### ISSUE THREAD

**Issue by Silver812** [DLR-RM/rl-baselines3-zoo#441]
### 🐛 Bug

I'm trying to use the Zoo to speed up the training and hyperparameter optimization of my custom environment. I'm still a beginner, so the environment is heavily based on [this example code](https://gymnasium.farama.org/main/tutorials/gymnasium_basics/environment_creation/#make-your-own-custom-environment) from the Farama Foundation (the main change is that I tried to convert it to a continuous observation space).

I have two problems:

**1. Training of my custom environment suddenly stops at 25000 timesteps.**
For some reason, the training can't go over that number. The 100K training does complete when using a prebuilt environment like CartPole though, so maybe it only happens with custom environments?
The problem persists when using different algorithms (I tried PPO and A2C).

**2. Optuna seems to initialize the optimization but then immediately exits.**
As I pasted in the relevant log output section, the initialization seems to go out well, but then immediately exits the process.

What I added in import_envs.py:
```
try:
    from rl_algorithm.envs.continuous_world import ContinuousWorldEnv
except ImportError:
    pass
```

What I added in ppo.yml (I copied the hyperparameters from another environment since I wasn't sure what to put here):
```
ContinuousWorld-v0:
  n_envs: 8
  n_timesteps: !!float 1e5
  policy: 'MultiInputPolicy'
  n_steps: 32
  batch_size: 256
  gae_lambda: 0.8
  gamma: 0.98
  n_epochs: 20
  ent_coef: 0.0
  learning_rate: lin_0.001
  clip_range: lin_0.2
```

I'm flagging this as a bug because I'm not getting any warnings or errors, it just weird behavior.

I can include the code of the custom environment in a reply if needed. 

### To Reproduce

```shell
python -m rl_zoo3.train --algo ppo --env ContinuousWorld-v0 --n-timesteps 100000 --progress
```
```shell
python -m rl_zoo3.train --algo ppo --env ContinuousWorld-v0 -n 50000 -optimize --n-trials 1000 --n-jobs 2 --sampler tpe --pruner median
```

### Relevant log output / Error message

Training problem last update before the 25000 timesteps mark:

```shell
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 1.15e+03 |
|    ep_rew_mean        | -4.4     |
| time/                 |          |
|    fps                | 3052     |
|    iterations         | 600      |
|    time_elapsed       | 7        |
|    total_timesteps    | 24000    |
| train/                |          |
|    entropy_loss       | -1.35    |
|    explained_variance | 0.0367   |
|    learning_rate      | 0.0007   |
|    n_updates          | 599      |
|    policy_loss        | 0.183    |
|    value_loss         | 0.156    |
------------------------------------
```

When close to 25000 it suddenly stops updating normally.

```shell
25% ━━━━━━━━━━━╺━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24,888/100,000  [ 0:00:13 < 0:00:24 , 3,166 it/s ]
```

After a few seconds of not updating, it does one final update exactly at 25000 timesteps and then stays there indefinitely without crashing.

```shell
25% ━━━━━━━━━━━╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 25,000/100,000  [ 0:00:32 < 0:01:00 , 1,260 it/s ]
```

Optimization problem log:

```shell          
2024-04-03 22:17:16.017613: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-04-03 22:17:16.627966: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
========== ContinuousWorld-v0 ==========
Seed: 3598421724
Loading hyperparameters from: C:\Users\silve\AppData\Local\Programs\Python\Python311\Lib\site-packages\rl_zoo3\hyperparams\ppo.yml
Default hyperparameters for environment (ones being tuned will be overridden):
OrderedDict([('batch_size', 256),
             ('clip_range', 'lin_0.2'),
             ('ent_coef', 0.0),
             ('gae_lambda', 0.8),
             ('gamma', 0.98),
             ('learning_rate', 'lin_0.001'),
             ('n_envs', 8),
             ('n_epochs', 20),
             ('n_steps', 32),
             ('n_timesteps', 100000.0),
             ('policy', 'MultiInputPolicy')])
Using 8 environments
Overwriting n_timesteps with n=50000
Doing 1 intermediate evaluations for pruning based on the number of timesteps. (1 evaluation every 100k timesteps)

Exiting...
```

### System Info

```
- OS: Windows-10-10.0.22631-SP0 10.0.22631 (It's actually Windows 11)
- Python: 3.11.4
- Stable-Baselines3: 2.3.0
- PyTorch: 2.0.1+cu118
- GPU Enabled: True
- Numpy: 1.24.3
- Cloudpickle: 2.2.1
- Gymnasium: 0.29.1
- SB3Zoo: 2.3.0
```

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/rl-baselines3-zoo/issues) in the repo
- [X] I have read the [SB3 documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] I have read the [RL Zoo documentation](https://rl-baselines3-zoo.readthedocs.io)
- [X] I have provided a [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014) example to reproduce the bug
- [X] I've used the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by Silver812** [comment]
Sorry, I didn't provide any code initially because it's way too long (assuming this was the checklist point I was missing). Here is my custom environment:

```
from enum import Enum
import sys
import numpy as np
import gymnasium as gym
from gymnasium import spaces
from stable_baselines3.common.vec_env import DummyVecEnv


class Actions(Enum):
    RIGHT = 0
    UP = 1
    LEFT = 2
    DOWN = 3


class ContinuousWorldEnv(gym.Env):
    metadata = {"render_modes": [None]}

    def __init__(self, size=50, render_mode=None):
        self.least_distance: np.inf
        self.last_action: 0
        self.size = size  # The size of the square grid
        self.window_size = 512  # The size of the PyGame window

        # Observations are dictionaries with the agent's and the target's location.
        # Each location is encoded as an element of {0, ..., `size`}^2, i.e. MultiDiscrete([size, size]).
        self.observation_space = spaces.Dict(
            # {
            #     "agent": spaces.Box(0, size - 1, shape=(2,), dtype=int),
            #     "target": spaces.Box(0, size - 1, shape=(2,), dtype=int),
            # }
            {
                "agent": spaces.Box(low=0, high=1, shape=(2,), dtype=np.float64),
                "target": spaces.Box(low=0, high=1, shape=(2,), dtype=np.float64),
            }
        )
        self._agent_location = np.array([-1, -1], dtype=int)
        self._target_location = np.array([-1, -1], dtype=int)

        # We have 4 actions, corresponding to "right", "up", "left", "down"
        self.action_space = spaces.Discrete(4)

        """
        The following dictionary maps abstract actions from `self.action_space` to
        the direction we will walk in if that action is taken.
        i.e. 0 corresponds to "right", 1 to "up" etc.
        """
        self._action_to_direction = {
            Actions.RIGHT.value: np.array([1, 0], dtype=int),
            Actions.UP.value: np.array([0, 1], dtype=int),
            Actions.LEFT.value: np.array([-1, 0], dtype=int),
            Actions.DOWN.value: np.array([0, -1], dtype=int),
        }

        """
        If human-rendering is used, `self.window` will be a reference
        to the window that we draw to. `self.clock` will be a clock that is used
        to ensure that the environment is rendered at the correct framerate in
        human-mode. They will remain `None` until human-mode is used for the
        first time.
        """
        self.window = None
        self.clock = None

    def _normalize_observation(self, observation, size):
        normalized_observation = {}

        for key, value in observation.items():
            normalized_observation[key] = value / (size - 1)

        return normalized_observation

    def _get_distance(self):
        return np.linalg.norm(self._agent_location - self._target_location, ord=1)

    def _get_obs(self):
        return {"agent": self._agent_location, "target": self._target_location}

    def _get_info(self):
        return {"distance": self._get_distance()}

    def reset(self, seed=None, options=None):
        # We need the following line to seed self.np_random
        super().reset(seed=seed)

        # Choose the agent's location uniformly at random
        self._agent_location = self.np_random.integers(0, self.size, size=2, dtype=int)

        # We will sample the target's location randomly until it does not coincide with the agent's location
        self._target_location = self._agent_location
        while np.array_equal(self._target_location, self._agent_location):
            self._target_location = self.np_random.integers(0, self.size, size=2, dtype=int)

        self.last_action = 0
        self.least_distance = self._get_distance()
        # observation = self._get_obs()
        observation = self._normalize_observation(self._get_obs(), self.size)
        info = self._get_info()

        return observation, info

    def _get_reward(self, action):

        distance = self._get_distance()
        reward = 0.0

        # If the distance to the target is less than 1, the agent has reached the target
        if distance < 1:
            print("Agent has reached the target")
            reward += 1.0

        # If the distance to the target is less than the least distance, the agent is getting closer to the target
        elif distance < self.least_distance:
            self.least_distance = distance
            # print("Agent is getting closer to the target")
            reward += 0.01

        # If the distance to the target is greater than the least distance, the agent is getting farther from the target
        else:
            reward -= 0.01

        if action != self.last_action:
            reward += 0.01
        else:
            reward -= 0.01

        return reward

    def step(self, action):
        # Map the action (element of {0,1,2,3}) to the direction we walk in
        direction = self._action_to_direction[action]
        # We use `np.clip` to make sure we don't leave the grid
        self._agent_location = np.clip(self._agent_location + direction, 0, self.size - 1)
        # An episode is done iff the agent has reached the target
        terminated = np.array_equal(self._agent_location, self._target_location)
        reward = self._get_reward(action)
        self.last_action = action
        # observation = self._get_obs()
        observation = self._normalize_observation(self._get_obs(), self.size)
        info = self._get_info()

        return observation, reward, terminated, False, info

    def close(self):
        sys.exit("\nExiting...\n")
```

**Comment by araffin** [comment]
> def close(self):
>        sys.exit("\nExiting...\n")

**Comment by Silver812** [comment]
Wow, that's a little embarrassing... Thanks for pointing it out though.

After changing that line to:
`print("Closing the environment")`

And adding a very simple truncating mechanism in the step function:
```
        truncated = False

        if self.timestep_counter >= 30000:
            truncated = True
            print("Episode truncated")
        else:
           self.timestep_counter += 1
```

The Optuna optimization starts normally, and it does seem to work alright. Just for good measure I'll add a partial log of the optimization procress:

```
========== ContinuousWorld-v0 ==========
Seed: 1257939581
Loading hyperparameters from: C:\Users\silve\AppData\Local\Programs\Python\Python311\Lib\site-packages\rl_zoo3\hyperparams\ppo.yml
Default hyperparameters for environment (ones being tuned will be overridden):
OrderedDict([('batch_size', 256),
             ('clip_range', 'lin_0.2'),
             ('ent_coef', 0.0),
             ('gae_lambda', 0.8),
             ('gamma', 0.98),
             ('learning_rate', 'lin_0.001'),
             ('n_envs', 8),
             ('n_epochs', 20),
             ('n_steps', 32),
             ('n_timesteps', 100000.0),
             ('policy', 'MultiInputPolicy')])
Using 8 environments
Overwriting n_timesteps with n=10000
Doing 1 intermediate evaluations for pruning based on the number of timesteps. (1 evaluation every 100k timesteps)
Closing the environment
Optimizing hyperparameters
C:\Users\silve\AppData\Local\Programs\Python\Python311\Lib\site-packages\optuna\samplers\_tpe\sampler.py:319: ExperimentalWarning: ``multivariate`` option is an experimental feature. The interface can change in the future.
  warnings.warn(
Sampler: tpe - Pruner: median
[I 2024-04-04 11:50:17,023] A new study created in memory with name: no-name-bcadd434-e207-47bc-9e99-3d0a25e03ea7
Agent has reached the target
Agent has reached the target
Episode truncated
Episode truncated
Episode truncated
Episode truncated
Episode truncated
Closing the environment
Closing the environment
Closing the environment
Closing the environment
Closing the environment
Closing the environment
Closing the environment
Closing the environment
Closing the environment
[I 2024-04-04 11:54:18,366] Trial 1 finished with value: -239.58799999999997 and parameters: {'batch_size': 16, 'n_steps': 64, 'gamma': 0.9999, 'learning_rate': 0.0003767443448677662, 'ent_coef': 0.0006887248484492611, 'clip_range': 0.4, 'n_epochs': 5, 'gae_lambda': 0.95, 'max_grad_norm': 0.6, 'vf_coef': 0.22057943399371072, 'net_arch': 'medium', 'activation_fn': 'relu'}. Best is trial 1 with value: -239.58799999999997.
Agent has reached the target
Episode truncated

... etc ...
```

As expected, the results are quite bad, so I'll be tweaking the timesteps and few other things.

On the first topic though, **the training hanging at 25000 timesteps**, it is still happening. What do you think is the problem?

And one last thing, I don't know if this is intended behavior, but if I want to stop the Optuna optimization process, it does not stop with the typical `Crtl + C` on the terminal. To truly stop it I have to kill the Python process from the Windows Task Manager.

**Comment by Silver812** [comment]
I am still dealing with the training problem. Can anybody point me in right direction?

**Comment by Silver812** [comment]
Hey, thanks for the reply qgallouedec. Yes, I have checked the environment with both the Gymnasium and the SB3 check_env() functions (I know one of them is a superset of the other, but I wanted to be sure). Neither of them found any issues. 

```
from gymnasium.utils.env_checker import check_env
from stable_baselines3.common.env_checker import check_env as sb3_check_env
```

**Comment by Silver812** [comment]
I finally solved the mystery. The training seemingly stopping at 25000 steps is to perform an evaluation. In fact, it's very easy to change this number by just using the "--eval-freq" argument.

Both of the problems that I mentioned here as bugs were from my own creation because I didn't even read the train.py source code or didn't properly check my own code before posting an issue.

**So I wanted to publicly apologize to @qgallouedec and specially @araffin for wasting their time on this.**

I'll try to be more careful when asking for help in the future.

By the way, I noticed this thanks to the incredible speed of the sbx implementation. It allowed me to wait past my impatience for enough time (just a few seconds) for the evaluation to finish and the counter to go above 25000. So yeah, thanks for making that awesome library too.
